### PR TITLE
[LS] Add returns keyword to function type desc context

### LINKED ISSUE #31196: [LS] returns keyword not suggested for function type descriptor
**Description:**
$subject encountered when a a function type descriptor is used as a function param in following scenarios.

**Steps to reproduce:**
Use the following code segments
```ballerina
function func1(string mystr, int myint) returns function (int,function (int,int) ret<cursor>)  {
    return myFunc;
}
```
```ballerina
function func2(int a, int b ,function(int,int) re<cursor>) {

}
```
**Affected Versions:**
slbeta2-SNAPSHOT

### PR DESCRIPTION
**Pull request by malinthar**
## Purpose
$subject to fix some of the specific scenarios depicted in the samples below.

Fixes #31196 

## Samples

sample 1

s1

sample2

s2

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by nadeeshaan — CHANGES_REQUESTED** [review]
We are handling this as a specific case. Without doing that, first, let's introduce ReqiuredParameterNodeContext and also, we need a parser level fix as well.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            /*
+              covers the following
+              function (int,int,..) re<cursor>
+            */
+            if (typeSymbol.isPresent() && typeSymbol.get().typeKind() == TypeDescKind.FUNCTION) {
+                Optional<TypeSymbol> returnType = ((FunctionTypeSymbol) typeSymbol.get()).returnTypeDescriptor();
+                if (returnType.isPresent() && returnType.get().typeKind() == TypeDescKind.NIL) {
+                    completionItems.add(new SnippetCompletionItem(context, Snippet.KW_RETURNS.get()));
+                }
+            } else {
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java:78]: We are taking this decision based on the semantic information. The best approach would be to make the decision based on the syntax nodes.

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java (hunk 1 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java]: Done

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        });
+    }
+
+    private boolean withinParamName(RequiredParameterNode node, BallerinaCompletionContext context) {
+        int cursor = context.getCursorPositionInTree();
+        Optional<Token> paramName = node.paramName();
+        if (paramName.isEmpty()) {
+            return false;
+        }
+        return paramName.get().textRange().startOffset() <= cursor && cursor <= paramName.get().textRange().endOffset();
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java:103]: Merge these

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                Optional<FunctionSignatureNode> functionSignature =
+                        ((FunctionTypeDescriptorNode) node.typeName()).functionSignature();
+                if (functionSignature.isPresent() && functionSignature.get().returnTypeDesc().isEmpty()) {
+                    completionItems.add(new SnippetCompletionItem(context, Snippet.KW_RETURNS.get()));
+                }
+            } else {
+                /*
+                Covers the Following
+                (1) function(T<cursor>)
+                */
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java:80]: With this approach, aren't we getting a hit to the else clause for the following case as well? 
`function (int a<cursor>)`

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java (hunk 3 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java:80]: Noted. We might get. Let me add a test case for that as well

Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java:80]: My concern is, we should not provide completions in that particular case. Currently we provide completions for that particular context right? If so shall we handle that as well?

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java (hunk 3 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java:80]: Have handled it.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+              function (int,int,..) re<cursor>
+            */
+            if (node.typeName() instanceof FunctionTypeDescriptorNode && withinParamName(node, context)) {
+                Optional<FunctionSignatureNode> functionSignature =
+                        ((FunctionTypeDescriptorNode) node.typeName()).functionSignature();
+                if (functionSignature.isPresent() && functionSignature.get().returnTypeDesc().isEmpty()) {
+                    completionItems.add(new SnippetCompletionItem(context, Snippet.KW_RETURNS.get()));
+                }
+            } else if (!withinParamName(node, context)) {
+                /*
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java:77]: Handles `int xxx<cursor>`

**Review by nadeeshaan — DISMISSED** [review]
LGTM

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Pull request merged**